pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -445,26 +448,58 @@ func applyUpstreamTLSSettings(cluster *v2.Cluster, tls *networking.TLSSettings)
 	case networking.TLSSettings_MUTUAL, networking.TLSSettings_ISTIO_MUTUAL:
 		cluster.TlsContext = &auth.UpstreamTlsContext{
 			CommonTlsContext: &auth.CommonTlsContext{
-				TlsCertificates: []*auth.TlsCertificate{
-					{
-						CertificateChain: &core.DataSource{
-							Specifier: &core.DataSource_Filename{
-								Filename: tls.ClientCertificate,
-							},
-						},
-						PrivateKey: &core.DataSource{
-							Specifier: &core.DataSource_Filename{
-								Filename: tls.PrivateKey,
-							},
-						},
-					},
-				},
 				ValidationContextType: &auth.CommonTlsContext_ValidationContext{
 					ValidationContext: certValidationContext,
 				},
 			},
 			Sni: tls.Sni,
 		}
+		if meshConfig.SdsUdsPath == "" || tls.Mode == networking.TLSSettings_MUTUAL {
+			cluster.TlsContext.CommonTlsContext.TlsCertificates = []*auth.TlsCertificate{
```

Comment:
nit: The `auth.TlsCertificate` and the following `auth.SdsSecretConfig` looks almost the same in here and authentication.go, maybe create 2 utility functions so that you can avoid the duplicate, could also be used in the test.

Reply:
thanks, done(not in unit test to avoid dev/test call into same codepath).